src/dune_rules/ocaml_stdlib.ml

```diff
@@ -12,20 +12,42 @@ let syntax =
     Dune_lang.Syntax.create
       ~name:"experimental_building_ocaml_compiler_with_dune" ~experimental:true
       ~desc:"experimental feature for building the compiler with dune"
-      [ ((0, 1), `Since (1, 3)) ]
+      [ ((0, 1), `Since (1, 3)); ((0, 2), `Since (3, 9)) ]
   in
   Dune_project.Extension.register_simple syntax (Dune_lang.Decoder.return []);
   syntax
 
+module Internal_modules = struct
+  let field ~loc:stanza_loc ~version =
+    let open Dune_lang.Decoder in
+    let old_field, field =
+      let field parser =
+        field "internal_modules" parser ~default:Predicate_lang.false_
+      in
+      let old_field = field (Glob.decode >>| Predicate_lang.Glob.of_glob)
+      and field =
+        field (repeat string >>| Predicate_lang.Glob.of_string_list)
+      in
+      (old_field, field)
+    in
+    let expected_version = (0, 2) in
+    if Syntax.Version.Infix.(version >= expected_version) then field
+    else
+      try_ old_field (fun _exn ->
```

Comment:
If the `internal_modules` parse fails, you should re-raise the old error (`exn`).

By the way, this feature is experimental so this versioning isn't really needed. The fields are backward compatible enough that we can just tell people to move on.